osu.Game/Overlays/OverlayScrollContainer.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Framework.Allocation;
+using osu.Framework.Extensions.Color4Extensions;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Effects;
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Input.Events;
+using osu.Game.Graphics.Containers;
+using osuTK;
+using osuTK.Graphics;
+
+namespace osu.Game.Overlays
+{
+    /// <summary>
+    /// <see cref="OsuScrollContainer"/> which provides <see cref="ScrollToTopButton"/>. Mostly used in <see cref="FullscreenOverlay"/>.
+    /// </summary>
+    public class OverlayScrollContainer : OsuScrollContainer
+    {
+        /// <summary>
+        /// Scroll position at which the <see cref="ScrollToTopButton"/> will be shown.
+        /// </summary>
+        private const int button_scroll_position = 200;
+
+        public ScrollToTopButton Button { get; }
+
+        private float currentTarget;
+
+        public OverlayScrollContainer()
+        {
+            AddInternal(Button = new ScrollToTopButton
+            {
+                Anchor = Anchor.BottomRight,
+                Origin = Anchor.BottomRight,
+                Margin = new MarginPadding(20),
+                Action = () =>
+                {
+                    ScrollToStart();
+                    currentTarget = Target;
+                    Button.State = Visibility.Hidden;
+                }
+            });
+        }
+
+        protected override void UpdateAfterChildren()
+        {
+            base.UpdateAfterChildren();
+
+            if (ScrollContent.DrawHeight + button_scroll_position < DrawHeight)
+            {
+                Button.State = Visibility.Hidden;
+                return;
+            }
+
+            if (Target == currentTarget)
```

Comment:
What's the reason for this `currentTarget` variable? As far as I can see it has no real purpose other than to maybe reduce amount of times the property is set? Prefer removal if that's the case.

Reply:
Clicking on button should cause it's immediately disappearance. Removing that will cause it only if we are close enugh to the start.

Reply:
And if we will leave everything as-is - setting button to Hidden state will be reverted on the next frame

Reply:
at very least that should be commented, since it is not immediately obvious.

Reply:
If that's true, it would already be the case. Because what you're saying implies `Target` is changing, which if it is will pass through that condition and cause the further comparison.

But that's not how `Target` works - it doesn't change all the time.

Am I misunderstanding?

Reply:
Tests pass with the early exit present or not. @EVAST9919 If it is required, please add comments and test coverage. I think you will require it's not and is left-over from a previous solution.